File: 79-word-search/word-search.cpp
class Solution {
public:
    bool backtracking(int i, int j, vector<vector<char>>& board, string word, int ROW, int COL, int index){
        if (index == word.length()) return true;
        if (0 > i || i >= ROW || 0 > j || j >= COL || board[i][j] != word[index]) return false;
        
        board[i][j] = '#';
        bool result = backtracking(i + 1, j, board, word, ROW, COL, index + 1) || backtracking(i - 1, j, board, word, ROW, COL, index + 1) || backtracking(i, j + 1, board, word, ROW, COL, index + 1) || backtracking(i, j - 1, board, word, ROW, COL, index + 1);
        board[i][j] = word[index];
        return result;


    }
    bool exist(vector<vector<char>>& board, string word) {
        int ROW = board.size();
        int COL = board[0].size();
        for (int i = 0; i < ROW; i++){
            for (int j = 0; j < COL; j++){
                if(backtracking(i, j, board, word, ROW, COL, 0)) return true;
            }
        }
        return false;
    }
};

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:

        ROW = len(board)
        COL = len(board[0])
        path = set()
        def backtracking(row, col, index):
            
            if index == len(word): return True
            if 0 > row or row >= ROW or 0 > col or col >= COL or board[row][col] != word[index] or (row, col) in path: return False

            index += 1
            path.add((row, col))
            res = backtracking(row + 1, col, index) or backtracking(row - 1, col, index) or backtracking(row, col + 1, index) or backtracking(row, col - 1, index)
            path.remove((row, col))
            return res

        for i in range(ROW):
            for j in range(COL):
                if backtracking(i, j, 0):
                    return True
        return False